Source/Tools/FEXLoader/LinuxSyscalls/SignalDelegator.cpp

```diff
@@ -1376,13 +1383,105 @@ namespace FEX::HLE {
 
   /**  @} */
 
+  static bool IsAsyncSignal(siginfo_t* Info, int Signal) {
+    if (Info->si_code <= SI_USER) {
+      // If the signal is not from the kernel then it is always async.
+      // This is because synchronous signals can be sent through tgkill,sigqueue and other methods.
+      // SI_USER == 0 and all negative si_code values come from the user.
+      return true;
+    }
+    else {
+      // If the signal is from the kernel then it is async only if it isn't an explicit synchronous signal.
+      switch (Signal) {
+        // These are all synchronous signals.
+        case SIGBUS:
+        case SIGFPE:
+        case SIGILL:
+        case SIGSEGV:
+        case SIGTRAP:
+          return false;
+        default: break;
+      }
+    }
+
+    // Everything else is async and can be deferred.
+    return true;
+  }
+
   void SignalDelegator::HandleGuestSignal(FEXCore::Core::InternalThreadState *Thread, int Signal, void *Info, void *UContext) {
+    ucontext_t* _context = (ucontext_t*)UContext;
+    auto SigInfo = *static_cast<siginfo_t*>(Info);
+
+    constexpr bool SupportDeferredSignals = true;
+    if (SupportDeferredSignals) {
+      auto RefCount = Thread->CurrentFrame->State.DeferredSignalRefCount.Load();
```

Comment:
Can make the intent a bit more explicit here:

```suggestion
      auto MustDeferSignal = (Thread->CurrentFrame->State.DeferredSignalRefCount.Load() != 0);
```